dpp/src/data_contract/state_transition/data_contract_update_transition/validation/basic/schema_compatibility_validator.rs

```diff
@@ -0,0 +1,384 @@
+/// The Schema compatibility validator is a port of a JavaScript version
+/// https://bitbucket.org/atlassian/json-schema-diff-validator/src/master/
+///
+/// The functionality has been ported 'as is' without any logic improvements and optimizations
+use std::convert::TryFrom;
+
+use anyhow::Context;
+use itertools::Itertools;
+use json_patch::PatchOperation;
+use jsonptr::{Pointer, Resolve};
+use serde_json::Value as JsonValue;
+use thiserror::Error;
+
+const PROPERTY_REQUIRED: &str = "required";
+const PROPERTY_DEFINITIONS: &str = "definitions";
+const PROPERTY_REF: &str = "$ref";
+
+#[derive(Default, Debug, Clone)]
+pub struct ValidationOptions {
+    pub allow_new_one_of: bool,
+    pub allow_new_enum_value: bool,
+    pub allow_reorder: bool,
+    pub deprecated_items: Vec<String>,
+}
+
+struct RemovedItem {
+    name: String,
+    operation: PatchOperation,
+}
+
+#[derive(Error, Debug)]
+pub enum DiffVAlidatorError {
+    /// The error that happens when there there is a problem with access to the fields
+    /// in dynamic data structure
+    #[error("Error while validation: {0}")]
+    DataStructureError(#[from] anyhow::Error),
+    /// The error is returned when validation proceeded but the schemas are not compatible
+    #[error("The schemas are not compatible: {diffs:#?}")]
+    SchemaCompatibilityError { diffs: Vec<PatchOperation> },
+}
+
+pub fn validate_schema_compatibility(
+    original_schema: &JsonValue,
+    new_schema: &JsonValue,
+) -> Result<(), DiffVAlidatorError> {
+    validate_schema_compatibility_with_options(
+        original_schema,
+        new_schema,
+        ValidationOptions::default(),
+    )
+}
+
+pub fn validate_schema_compatibility_with_options(
+    original_schema: &JsonValue,
+    new_schema: &JsonValue,
+    opts: ValidationOptions,
+) -> Result<(), DiffVAlidatorError> {
+    let patch = json_patch::diff(original_schema, new_schema);
+    let mut diff: Vec<PatchOperation> = vec![];
+    let mut removed: Vec<RemovedItem> = vec![];
+    let mut inserted: Vec<String> = vec![];
+
+    for operation in patch.0.iter() {
+        match operation {
+            PatchOperation::Move(op_move) => {
+                if is_operation_move_remove_compatible(
+                    op_move.path.as_str(),
+                    original_schema,
+                    &opts,
+                )? {
+                    continue;
+                }
+                diff.push(operation.clone());
```

Comment:
Just wondering - why not use `into_iter` and avoid cloning

Reply:
Actually, in the second  I got rid of cloning and I also did some small refactoring